ggml/src/ggml.c

```diff
@@ -2995,42 +3114,36 @@ inline static void ggml_critical_section_start(void) {
 }
 
 #ifdef GGML_USE_OPENMP
-static void ggml_barrier(struct ggml_compute_state_shared * shared) {
-    if (shared->n_threads == 1) {
+static void ggml_barrier(struct ggml_compute_threadpool * threadpool) {
+    if (threadpool->n_threads_cur == 1) {
         return;
     }
 
     #pragma omp barrier
 }
 #else
-static void ggml_barrier(struct ggml_compute_state_shared * shared) {
-    if (shared->n_threads == 1) {
+static void ggml_barrier(struct ggml_compute_threadpool * threadpool) {
+    if (threadpool->n_threads_cur == 1) {
         return;
     }
 
-    atomic_int * n_barrier = &shared->n_barrier;
-    atomic_int * n_barrier_passed = &shared->n_barrier_passed;
+    atomic_int * n_barrier = &threadpool->n_barrier;
+    atomic_int * n_barrier_passed = &threadpool->n_barrier_passed;
 
-    int n_threads = shared->n_threads;
-    int passed_old = atomic_load(n_barrier_passed);
+    int n_threads = threadpool->n_threads_cur;
+    int passed_old = atomic_load_explicit(n_barrier_passed, memory_order_relaxed);
 
     if (atomic_fetch_add(n_barrier, 1) == n_threads - 1) {
         // last thread
         atomic_store(n_barrier, 0);
-        atomic_fetch_add(n_barrier_passed, 1);
+        atomic_fetch_add_explicit(n_barrier_passed, 1, memory_order_relaxed);
     } else {
         // wait for other threads
-        const int n_spin_before_sleep = 100000;
         while (true) {
-            for (int i = 0; i < n_spin_before_sleep; i++) {
-                if (atomic_load(n_barrier_passed) != passed_old) {
-                    return;
-                }
-            #if defined(__SSE3__)
-                _mm_pause();
-            #endif
+            if (atomic_load_explicit(n_barrier_passed, memory_order_relaxed) != passed_old) {
+                return;
             }
-            sched_yield();
+            __cpu_relax();
```

Comment:
It depends :) libgomp (gcc default) doesn't call sched_yeld() but does block on mutex depending on the `OMP_WAIT_POLICY`. OMP_WAIT_POLICY on my Ryzen9 system is set to active. In that mode GOMP threads spin without yield or mutex.
libomp (clang default) does lots of sched_yeld() and then sleep on mutex, again depending on the OMP_WAIT_POLICY.

I did lots of profiling earlier and noticed that threadpool had much higher context switch counts than GOMP.
One of the sources was sched_yeld() (and the other was original pause/resume which I improved since then).

sched_yield() here (in the ggml_barrier) in general is a not a good idea. It only creates extra load on the system (ie syscalls and context switches). We want the barriers to be as lean as possible since they are quire frequent and short lived. So it makes sense to just spin and use the instruction that puts CPU in a less power hungry state.
Of course it makes perfect sense to sleep between the calls to graph_compute though, which is what we do now with the hybrid polling (ie poll then sleep).